### DISCUSSION TITLE
How to setup a password for redis inside the Rocket.toml file?

### DISCUSSION THREAD

**Discussion by patchthecode** [rwf2/Rocket#2331] [Questions]
I swear this is the hardest thing to find on the internet.

My Redis instance has a password.
How do i configure this from an environment variable in the `Rocket.toml` file?

Here is what i have inside `Rocket.toml`. And it works perfectly if redis has no password.

```
[release.databases.redis]
url = "redis://redis_trending_server"
```

`redis_trending_server` here is my redis server given from docker-compose file. This works. But now I have set a redis password.
I have also passed this password in an environment variable to Rust ${MY_REDIS_PASSWORD}

Now inside the `Rocket.toml` file how do i first
1) Grab that environment variable
and then 
2) configure that password in the url above?

I cant find a single documentation any where on this -__-

**Comment by blackghost1987** [comment]
I haven't actually used redis with Rocket before, but the idea should not be much more different with other database engines.

To put the password in the database URI, you'll have to use the following format:
```
redis://username:password@address
```

The problem is you cannot really "append" the password to the URL in the Rocket.toml, it needs the complete URL with the password in it. Instead you have to overwrite the whole value with a new one, containing the password as well. To do that you'll need to set the following env var:
```
ROCKET_databases.redis.url
```

Yes, the name of the Env var has lowercase letters and dots as well, which is kinda strange and some shells don't like it, but this is how it works for now unfortunately.

If you only have the password in the MY_REDIS_PASSWORD env var, then you'll have to come up with some Docker magic to concatenate it to the URL and put it into the Rocket specific env var.

**Reply by edgarogh** [reply]
"Detaching" the password from the database URL is quite uncommon from my experience. I'd personally just set the entire database URL at once from the host docker CLI in this situation.

And as you said, other options can include Docker magic, but it's also possible to implement the feature for providing passwords separately in OP's crate.

The rocket guide explains how to [programmatically override some parameters](https://rocket.rs/v0.5-rc/guide/configuration/#custom-providers) (in OP's case, it would be `databases { redis { url } }`) or how to [extract (potentially custom) configuration](https://rocket.rs/v0.5-rc/guide/configuration/#extracting-values). For instance, it should be possible to extract the specified "database URL" _without the password_, and merge it with another "database password" configuration parameter that would be provided separately (with a URI parsing/building crate).

**Reply by patchthecode** [reply]
I just checked out that documentation.. but it doesnt say how to get an env variable into the toml file.

where exactly do i put this? `databases { redis { url } })`

**Reply by edgarogh** [reply]
Sorry, I explained too fast. `databases { redis { url } }` was just pseudo-code for a serde model. Basically, Rocket uses a library called [figment](https://docs.rs/figment/latest/figment/) for reading config values. _figment_ is able to extract from `Rocket.toml`, but also from environment variables (and more), all automatically. If you follow the guide paragraphs I've sent, you will freely get the deserialization of environment variables in addition to new Rocket.toml keys.
